Xamarin.Build.AsyncTask/BackgroundTaskManager.cs

```diff
@@ -0,0 +1,95 @@
+    {
+
+        ConcurrentDictionary<string, ConcurrentBag<TPL.Task>> tasks = new ConcurrentDictionary<string, ConcurrentBag<TPL.Task>> ();
+        CancellationTokenSource tcs = new CancellationTokenSource ();
+
+        /// <summary>
+        /// Get an Instance of the TaskManager.
+        /// NOTE This MUST be called from the main thread in a Task, it cannot be called on a background thread.
+        /// </summary>
+        /// <param name="buildEngine4">An instance of the IBuildEngine4 interface.</param>
+        /// <returns>An instance of a TaskManager</returns>
+        public static BackgroundTaskManager GetTaskManager (IBuildEngine4 buildEngine4)
+        {
+            var manager = (BackgroundTaskManager)buildEngine4.GetRegisteredTaskObject (typeof (BackgroundTaskManager).FullName, RegisteredTaskObjectLifetime.Build);
+            if (manager == null)
+            {
+                manager = new BackgroundTaskManager ();
+                buildEngine4.RegisterTaskObject (typeof (BackgroundTaskManager).FullName, manager, RegisteredTaskObjectLifetime.Build, allowEarlyCollection: false);
+            }
+            return manager;
+        }
+
+        /// <summary>
+        /// Register a new Task which will be running in the background.
+        /// If you have multiple tasks running you can split them up into
+        /// different categories. This can then we used to wait in differnt
+        /// parts of the build later on.
+        /// </summary>
+        /// <param name="task">The task you are running</param>
+        /// <param name="category">The category this task is in. </param>
+        public void RegisterTask (TPL.Task task, string category = "default")
+        {
+            var bag = tasks.GetOrAdd (category, new ConcurrentBag<TPL.Task> ());
+            bag.Add (task);
+        }
+
+        /// <summary>
+        /// Returns an array of Tasks for that category.
+        /// </summary>
+        /// <param name="category">The category you want to get the list for.</param>
+        /// <returns>Either the array of task or an empty array if the category does not exist.</returns>
+        public TPL.Task [] this [string category]
+        {
+            get
+            {
+                if (!tasks.ContainsKey (category))
+                    return new TPL.Task [0];
+                return tasks [category].ToArray ();
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of tasks for a particular category.
+        /// </summary>
+        /// <param name="category">The category you want to get the count for.</param>
+        /// <returns>An integer with the number of registered taks for that category or zero</returns>
+        public int GetCategoryCount (string category)
+        {
+            if (!tasks.ContainsKey (category))
+                return 0;
+            return tasks [category].Count;
+        }
+
+        public void Dispose ()
+        {
+            tcs.Cancel ();
```

Comment:
I wonder if the dispose should wait on all tasks, so that you can more easily use this by just queueing tasks and then assume that before the build is done, they will be automatically waited.... 
If someone has already waited on them, it would be a no-op anyway...

Reply:
I am adding this.